backend/alternative_flights/alt_flight_search.py

```diff
@@ -0,0 +1,232 @@
+
+            dfs(flight.arrival_airport, flight.arrival_time, new_option, new_visited)
+
+    initial_option = FlightOption()
+    dfs(departure_airport, departure_time, initial_option, {departure_airport})
+
+    return options
+
+def calculate_baseline_duration(options: List[FlightOption]) -> float:
+    direct_flights = [opt for opt in options if len(opt.flights) == 1]
+    if direct_flights:
+        return min(opt.total_duration for opt in direct_flights)
+   
+    min_connections = min(len(opt.flights) for opt in options)
+    least_connection_flights = [opt for opt in options if len(opt.flights) == min_connections]
+    return min(opt.total_duration for opt in least_connection_flights)
+
+def apply_discounts(options: List[FlightOption], baseline_duration: float):
+    durations = np.array([opt.total_duration for opt in options])
+    idle_times = np.array([opt.idle_time for opt in options])
+
+    duration_threshold = baseline_duration * 1.5
+    max_idle_time = np.max(idle_times)
+
+    for i, option in enumerate(options):
+        discount = 0
+        if durations[i] >= duration_threshold:
+            discount += option.total_price * 0.5
+            if idle_times[i] == max_idle_time and max_idle_time > 0:
+                discount += option.total_price * 0.2
+       
+        option.discount = min(discount, option.total_price * 0.7)  # Cap discount at 70%
+        option.total_price -= option.discount
+
+def calculate_ranking_scores(options: List[FlightOption], preferred_airline_code: str) -> List[FlightOption]:
+    prices = np.array([opt.total_price for opt in options])
+    durations = np.array([opt.total_duration for opt in options])
+    connections = np.array([len(opt.flights) - 1 for opt in options])
+    airline_matches = np.array([1 if any(flight.airline_code == preferred_airline_code for flight in opt.flights) else 0 for opt in options])
+
+    price_scores = 1 - (prices - np.min(prices)) / (np.max(prices) - np.min(prices))
+    speed_scores = 1 - (durations - np.min(durations)) / (np.max(durations) - np.min(durations))
+    directness_scores = np.where(connections == 0, 2, 1 - connections / np.max(connections))
+
+    overall_scores = (
+        0.35 * price_scores +
+        0.25 * speed_scores +
+        0.30 * directness_scores +
+        0.10 * airline_matches
+    )
+
+    scored_options = list(zip(overall_scores, options))
+    scored_options.sort(key=lambda x: (-x[0], x[1].total_price))
+    sorted_options = [option for _, option in scored_options]
+
+    for score, option in zip(overall_scores, sorted_options):
+        option.score = score
+
+    return sorted_options
+
+def main():
+    current_dir = os.path.dirname(os.path.abspath(__file__))
+    file_path = os.path.join(current_dir, 'flight_data.json')
+   
+    flight_data = load_flight_data(file_path)
+
+    departure_airport = "NYC"
+    arrival_airport = "MIA"
+    departure_time = datetime(2024, 7, 10, 12, 0, 0)
+    preferred_airline_code = "AA"
```

Comment:
You will need to take those as parameter while consuming this code from express layer, right?

Reply:
Yeap. Working on it now